### ISSUE TITLE
Using the option ID for dropdown ID's instead of value.

### ISSUE THREAD

**Issue by timvanuum** [DashboardCode/BsMultiSelect#16]
Hi there,

First of all great plugin.

But we encountered a little problem.
We're using parsley for some validation.
And it conflicts with the ID's containing space and other chars like ( and ).

I've added a solution in, but it would be great if it would be implemented.
On line 295 form BsMultiSelect.js I added: 
`` var optionId = typeof(optionElement.id) !== 'undefined' && optionElement.id.length > 0 ?  optionElement.id : optionElement.value;``

And changed:
``var adjustDropDownItem = this.adapter.CreateDropDownItemContent($dropDownItem, optionElement.value, itemText);``

to

``var adjustDropDownItem = this.adapter.CreateDropDownItemContent($dropDownItem, optionId, itemText);``

**Comment by rpokrovskij** [comment]
Thank you, very much. 
Can you clone a jsfiddle https://jsfiddle.net/u2xf6bew/3/ to demonstrate a problem? I do not remember that plugin would depend on id attribute (it generates some id for new DOM elements this is true)...

**Comment by timvanuum** [comment]
I've created a clone: https://jsfiddle.net/fmuL0gk9/

Here a screenshot of can cause trouble (in case of parsley for example).
https://ibb.co/8MxVzXc

With the code I've placed you prefer ID if it's set on the option, else it use the value.
What also be possible is to slug the value.

**Comment by rpokrovskij** [comment]
Got it: you are creating "predefinined" ids for options that can be reused as a part of checkbox id. This is ok, but I stack comparing your way to solve it with two other possibilities other options: escape "illegal" value chars and provide the special function through the API (which creates ids). In second case how the default function should look...

**Comment by rpokrovskij** [comment]
It is interesting that by specification ID and NAME both not allowed to contain spaces. From HTML4:  ID and NAME and tokens must begin with a letter ([A-Za-z]) and may be followed by any number of letters, digits ([0-9]), hyphens ("-"), underscores ("_"), colons (":"), and periods (".").

**Comment by rpokrovskij** [comment]
So the situation: you use invalid (HTML4/5) names with spaces, when I utilize them to create ids - I assign invalid ids... From this point of view component should act this way: it should provide to the user possibility to create the custom "drop down item" internals - with full customization user will get a possibility to create custom id (or not create at all, since those ids are not vital: currently they are used to ref label and checkbox and that is all).

**Comment by rpokrovskij** [comment]
I have just released version 0.3.0 where all internals are published (to be customizable) on configuration object. You can create your own id this way:

````
$('#bsMultiSelectJson').bsMultiSelect(
                        {
                             configuration: {
                                  createCheckBoxId : function(c,option){
                                                  return `${c.containerClass}-bsMultiSelectJson-generated-checkbox-${option.id}-id`;
                                              }
                             },
                                  createInputId: function(c){
                                                  return `${c.containerClass}-bsMultiSelectJson-filterInput-id`;
                                              }
                             },
                        }
                );
````
